lib/galaxy/tools/__init__.py

```diff
@@ -2185,6 +2184,13 @@ def handle_input(
             collection_info=collection_info,
             completed_jobs=completed_jobs,
         )
+
+        tags = incoming.get("tags", [])
+        if isinstance(tags, list) and tags:
+            tag_handler = self.app.tag_handler
+            for _, hda in execution_tracker.output_datasets:
+                tag_handler.apply_item_tags(user=trans.user, item=hda, tags_str=",".join(tags))
```

Comment:
```suggestion
                tag_handler.apply_item_tags(user=trans.user, item=hda, tags_str=",".join(tags), flush=False)
```
can't flush on each output here